zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
+                // anticipation of the chain verifying. But the chain is
+                // incomplete, so we have to put them back in the queue.
+                //
+                // The order here shouldn't matter, but add the blocks in
+                // height order, for consistency.
+                for vblock in rev_valid_blocks.drain(..).rev() {
+                    let height = vblock
+                        .block
+                        .coinbase_height()
+                        .expect("queued blocks have a block height");
+                    self.queued.entry(height).or_default().push(vblock);
+                }
+
+                // Ensure that we're making progress
+                assert_eq!(
+                    self.previous_checkpoint_height(),
+                    old_prev_check_height,
+                    "processing must not change progress on failure"
+                );
+                let current_target = self.target_checkpoint_height();
+                assert!(
+                    current_target == WaitingForBlocks
+                        || current_target < Checkpoint(target_checkpoint_height),
+                    "processing must decrease or eliminate target on failure"
+                );
+
+                // Stop verifying, and wait for the next valid block
+                return;
+            }
+        }
+
+        // The checkpoint and the parent hash must match.
+        // See the detailed checkpoint comparison comment above.
+        assert_eq!(
+            expected_hash, previous_checkpoint_hash,
+            "the previous checkpoint should match: bad checkpoint list, zebra bug, or bad chain"
+        );
+
+        // All the blocks we've kept are valid, so let's verify them
+        // in height order.
+        for qblock in rev_valid_blocks.drain(..).rev() {
+            // Sending can fail, but there's nothing we can do about it.
+            let _ = qblock.tx.send(Ok(qblock.hash));
+        }
+
+        // Finally, update the checkpoint bounds
+        self.update_progress(target_checkpoint_height);
+
+        // Ensure that we're making progress
+        assert!(
+            self.previous_checkpoint_height() > old_prev_check_height,
+            "progress must increase on success"
+        );
+        // We've cleared the target range, so this check should be cheap
+        let current_target = self.target_checkpoint_height();
+        assert!(
+            current_target == WaitingForBlocks || current_target == FinishedVerifying,
+            "processing must clear target on success"
+        );
+    }
+}
+
+/// CheckpointVerifier rejects pending futures on drop.
+impl Drop for CheckpointVerifier {
+    /// Send an error on `tx` for any `QueuedBlock`s that haven't been verified.
+    ///
+    /// We can't implement `Drop` on QueuedBlock, because `send()` consumes
+    /// `tx`. And `tx` doesn't implement `Copy` or `Default` (for `take()`).
```

Comment:
You might be able to work around this by storing `tx` in an `Option`, then you can do `Option::take` which will leave the member as `None`.

Reply:
Seems fine for `QueuedBlock` not to implement `Drop`.

Reply:
I don't think complicating the interface and code is worth this change.

I also think Drop is part of the contract of `CheckpointVerifier`, not `QueuedBlock`. Although the difference is unlikely to be significant. (And in general, there is no guarantee that all futures resolve.)